pkg/pool-manager/virtualmachine_pool.go

```diff
@@ -333,6 +355,124 @@ func (p *PoolManager) revertAllocationOnVm(vmName string, allocations map[string
+	configMap, err := p.kubeClient.CoreV1().ConfigMaps(names.MANAGER_NAMESPACE).Get(vmWaitConfigMapName, metav1.GetOptions{})
+	if err != nil {
+		if errors.IsNotFound(err) {
+			_, err = p.kubeClient.CoreV1().
+				ConfigMaps(names.MANAGER_NAMESPACE).
+				Create(&corev1.ConfigMap{ObjectMeta: metav1.ObjectMeta{Name: vmWaitConfigMapName,
+					Namespace: names.MANAGER_NAMESPACE}})
+
+			return map[string]string{}, nil
+		}
+
+		return nil, err
+	}
+
+	return configMap.Data, nil
+}
+
+// Add all the allocated mac addresses to the waiting config map with the current time.
+func (p *PoolManager) AddMacToWaitingConfig(allocations map[string]string) error {
+	configMap, err := p.kubeClient.CoreV1().ConfigMaps(names.MANAGER_NAMESPACE).Get(vmWaitConfigMapName, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+
+	if configMap.Data == nil {
+		configMap.Data = map[string]string{}
+	}
+
+	for _, macAddress := range allocations {
+		log.V(1).Info("add mac address to waiting config", "macAddress", macAddress)
+		macAddress = strings.Replace(macAddress, ":", "-", 5)
+		configMap.Data[macAddress] = time.Now().Format(time.RFC3339)
+	}
+
+	_, err = p.kubeClient.CoreV1().ConfigMaps(names.MANAGER_NAMESPACE).Update(configMap)
+	return err
+}
+
+// Remove all the mac addresses from the waiting configmap this mean the vm was saved in the etcd and pass validations
+func (p *PoolManager) MarkVMAsReady(vm *kubevirt.VirtualMachine) error {
+	p.poolMutex.Lock()
+	defer p.poolMutex.Unlock()
+
+	configMap, err := p.kubeClient.CoreV1().ConfigMaps(names.MANAGER_NAMESPACE).Get(vmWaitConfigMapName, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+
+	if configMap.Data == nil {
+		log.Info("the configMap is empty")
+		return nil
+	}
+
+	for _, vmInterface := range vm.Spec.Template.Spec.Domain.Devices.Interfaces {
+		if vmInterface.MacAddress != "" {
+			p.macPoolMap[vmInterface.MacAddress] = AllocationStatusAllocated
+			macAddress := strings.Replace(vmInterface.MacAddress, ":", "-", 5)
+			delete(configMap.Data, macAddress)
+		}
+	}
+
+	_, err = p.kubeClient.CoreV1().ConfigMaps(names.MANAGER_NAMESPACE).Update(configMap)
+	log.V(1).Info("marked virtual machine as ready", "virtualMachineNamespace", vm.Namespace,
+		"virtualMachineName", vm.Name)
+	return err
+}
+
+// This function check if there are virtual machines that hits the create
+// mutating webhook but we didn't get the creation event in the controller loop
+// this mean the create was failed by some other mutating or validating webhook
+// so we release the virtual machine
+func (p *PoolManager) vmWaitingCleanupLook(waitTime int) {
+	c := time.Tick(3 * time.Second)
+	log.Info("starting cleanup loop for waiting mac addresses")
+	for _ = range c {
+		p.poolMutex.Lock()
+
+		configMap, err := p.kubeClient.CoreV1().ConfigMaps(names.MANAGER_NAMESPACE).Get(vmWaitConfigMapName, metav1.GetOptions{})
+		if err != nil {
+			// TODO: should we exit here?
```

Comment:
Do we need to resolve this?

Reply:
we can remove this.

The next loop will try again